lib/fluent/supervisor.rb

```diff
@@ -346,64 +372,73 @@ def stop_rpc_server
       @rpc_server.shutdown
     end
 
-    def supervise(&block)
-      start_time = Time.now
-
-      Process.setproctitle("supervisor:#{@process_name}") if @process_name
+    def supervise
       $log.info "starting fluentd-#{Fluent::VERSION}"
 
-      if !Fluent.windows?
-        @main_pid = fork do
-          main_process(&block)
-        end
+      if ServerEngine.windows?
+        fluentd_spawn_cmd = ServerEngine.ruby_bin_path + ' ' + $0 + ' '
       else
-        if @supervise
-          fluentd_spawn_cmd = @rubybin_dir+"/ruby.exe '"+@rubybin_dir+"/fluentd' "
-          $fluentdargv.each{|a|
-            fluentd_spawn_cmd << (a + " ")
-          }
-          fluentd_spawn_cmd << ("--no-supervisor")
-          $log.info "spawn command to main (windows) : " + fluentd_spawn_cmd
-          @main_pid = Process.spawn(fluentd_spawn_cmd)
-        else
-          main_process(&block)
-        end
-      end
-
-      if @daemonize && @wait_daemonize_pipe_w
-        STDIN.reopen("/dev/null")
-        STDOUT.reopen("/dev/null", "w")
-        STDERR.reopen("/dev/null", "w")
-        @wait_daemonize_pipe_w.close
-        @wait_daemonize_pipe_w = nil
+        fluentd_spawn_cmd = $0 + ' '
       end
 
-      Process.waitpid(@main_pid)
-      @main_pid = nil
-      ecode = $?.to_i
+      $fluentdargv.each{|a|
+        fluentd_spawn_cmd << (a + " ")
```

Comment:
Is this cmd string executed vis shell?
If so, these should be escaped with ShellWords. (I don't know well about Windows though.)